filament/src/fg2/details/DependencyGraph.h

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (C) 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef TNT_FILAMENT_FG2_GRAPH_H
+#define TNT_FILAMENT_FG2_GRAPH_H
+
+#include <utils/ostream.h>
+
+#include <vector>
+
+namespace filament::fg2 {
+
+/**
+ * A very simple dependency graph (DAG) class that support culling of unused nodes
+ */
+class DependencyGraph {
+public:
+    DependencyGraph() noexcept;
+    DependencyGraph(const DependencyGraph&) noexcept = delete;
+    DependencyGraph& operator=(const DependencyGraph&) noexcept = delete;
+
+    using NodeID = uint32_t;
+
+    class Node;
+
+    /**
+     * A link between two nodes
+     */
+    class Edge {
```

Comment:
This might look a little cleaner if it were a `struct`.  Technically this merely lets you remove the public keywords (here, as well as two places in the descendant) but it also serves as a hint to humans: "hey, edges are merely holding some data, they do not have any logic"

Reply:
Good idea